wordpress-mobile/WordPress-Android
# IA Reader self-hosted improvements

Replaces and supersedes #10986 (point 1 was fixed in #11206)

This PR introduces enhancements to the new IA reader for users that have self-hosted sites only or they do have also WP.com account but currently logged off from it. (cc @osullivanchris )

**Note for the reviewer:** the PR dimension/scope should be still good enough I hope, but if you feel it's too large/wide let me know and will try to split in some way 🙇‍♂️.

What we introduced:

- When a user is logged out from WP.com and has at least a self-hosted site in the app, the reader is as follows:

| Id | Description | Image |
|---|---|:-:|
| Fig1 | On fresh install the Discover page is presented first (both wp.com and self-hosted) | ![image](https://user-images.githubusercontent.com/47797566/74345211-07478f80-4dae-11ea-9c73-716e03840243.png) |
| Fig2 | 3 tabs and `Filter` subfiltering + Empty view with dedicated message + no access to settings cog (nor search icon but this was not introduced in this PR) | ![image](https://user-images.githubusercontent.com/47797566/74343424-54763200-4dab-11ea-960f-0b02bfffe1af.png) |
| Fig3 | Empty SITES in bottomsheet with CTA to login; if tapped on CTA, user is landed to Me page to login | ![image](https://user-images.githubusercontent.com/47797566/74343656-b20a7e80-4dab-11ea-829e-f8a8ef73e1cb.png) |

## Notes
1. On user id switches or when the user passes from having access token (wp.com user) to not having it (pure self-hosted) and viceversa, we trigger a Blogs/Tags auto refresh when the user comes to the reader
2. As a follow up to #11206 we implemented an in memory default tag used as a sort of `null object` in cases where the tag table could be empty (see `getDbOrInMemoryDefaultTag ` function)
3. We added a `TagUpdateClientUtilsProvider` class to create a single point where to get the client for tags update

## To test
### Test case 1
- Make a fresh install and login with WP.com (no self-hosted)
- Go to Reader and check you are landed in the Discover page
- Go to `FOLLOWING` tab
- Logout from WP.com and add a self-hosted site
- Go to Reader and check you are landed in the Discover page as well (Fig1)
- Go to `Following` and check you get something similar to Fig2

### Test case 2
- Verify you are logged out from WP.com account and you have added a pure self-hosted site
- In `FOLLOWING` open bottom sheet and check you get something similar to Fig3
- Press the Login CTA and check you are landed into `Me` page
- Login into WP.com and come back to Reader
- Check you have now your followed SITES and TAGS similar to below image

<p align="center">
  <img width="300" src="https://user-images.githubusercontent.com/47797566/74348879-ade25f00-4db3-11ea-88c9-134b84aadb41.png">
</p>

### Test case 3
Smoke test the reader both with wp.com and self-hosted sites (below some of the possible steps for smoke testing but feel free to add 😊):
- Move between tabs
- Use the subfilter to select different sources (Sites/Tags) to navigate
- Open posts from the reader
- Save posts for later and open them from the `SAVED` tab (both WP.com and self-hosted)
- Follow/Unfollow tags/sites from the Reader (only for WP.com)
- Follow/Unfollow tags/sites from the Setting (only for WP.com)
- Whatever else you can think of 😊...

### Test case 4
See comment [here](https://github.com/wordpress-mobile/WordPress-Android/pull/11270#issuecomment-587904945)

**NOTE**: for the modifications 2 and 3 reported in the NOTES above, a part the smoke test I retested the steps reported in #11206 (good if you could also retry those steps, thanks).

PR submission checklist:
